src/Components/Server/src/Circuits/DefaultPersistedCircuitCache.cs

```diff
@@ -0,0 +1,169 @@
+        var options = new MemoryCacheEntryOptions
+        {
+            Size = 1,
+            PostEvictionCallbacks = { PostEvictionCallback },
+            ExpirationTokens = { new CancellationChangeToken(cancellationTokenSource.Token) },
+        };
+
+        var persistedCircuitEntry = new PersistedCircuitEntry
+        {
+            State = persistedCircuitState,
+            TokenSource = cancellationTokenSource
+        };
+
+        _persistedCircuits.Set(circuitId.Secret, persistedCircuitEntry, options);
+    }
+
+    private void OnEntryEvicted(object key, object value, EvictionReason reason, object state)
+    {
+        switch (reason)
+        {
+            case EvictionReason.Expired:
+            case EvictionReason.TokenExpired:
+            // Happens after the circuit state times out, this is triggered by the CancellationTokenSource we register
+            // with the entry, which is what controls the expiration
+            case EvictionReason.Capacity:
+                // Happens when the cache is full
+                var persistedCircuitEntry = (PersistedCircuitEntry)value;
+                Log.CircuitStateEvicted(_logger, persistedCircuitEntry.CircuitId, reason);
+                break;
+
+            case EvictionReason.Removed:
+                // Happens when the entry is explicitly removed as part of resuming a circuit.
+                return;
+            default:
+                Debug.Fail($"Unexpected {nameof(EvictionReason)} {reason}");
+                break;
+        }
+    }
+
+    public Task<PersistedCircuitState> RestoreCircuitAsync(CircuitId circuitId, CancellationToken cancellation = default)
+    {
+        Log.CircuitResumeStarted(_logger, circuitId);
+
+        lock (_lock)
+        {
+            var state = RestoreCore(circuitId);
+            if (state == null)
+            {
+                Log.FailedToFindCircuitState(_logger, circuitId);
+                return _noMatch;
+            }
+
+            return Task.FromResult(state);
+        }
+    }
+
+    private PersistedCircuitState RestoreCore(CircuitId circuitId)
+    {
+        if (_persistedCircuits.TryGetValue(circuitId.Secret, out var value) && value is PersistedCircuitEntry entry)
+        {
+            DisposeTokenSource(entry);
+            _persistedCircuits.Remove(circuitId.Secret);
+            Log.CircuitStateFound(_logger, circuitId);
+            return entry.State;
+        }
+
+        return null;
+    }
+
+    private void DisposeTokenSource(PersistedCircuitEntry entry)
+    {
+        try
+        {
+            entry.TokenSource.Dispose();
+        }
+        catch (Exception ex)
+        {
+            Log.ExceptionDisposingTokenSource(_logger, ex);
+        }
+    }
+
+    private class PersistedCircuitEntry
+    {
+        public PersistedCircuitState State { get; set; }
+
+        public CancellationTokenSource TokenSource { get; set; }
+
+        public CircuitId CircuitId { get; set; }
```

Comment:
I don't see this `CircuitId` property being set anywhere. Maybe that's missing from `PersistCore()`?

Reply:
Yeah, good catch. It was only used for logging, so that's why it went unnoticed.